src/pspm_convert_ppg2hb.m

```diff
@@ -12,162 +12,227 @@
-if length(pis)==1,warning('ID:OnePulse', 'Only one pulse found, unable to calculate min_pulse_period.');return;end
-
-% get pulse period lower limit (assumed onset) as 30% of smalest period
-% before detected peaks
-min_pulse_period = min(diff(pis));
-period_index_lower_bound = floor(pis(2:end-1)-.3*min_pulse_period);
-fprintf('...');
-
-% Create template from mean of peak time-locked ppg pulse periods
-pulses = cell2mat(arrayfun(@(x) data.data(x:x+min_pulse_period),period_index_lower_bound','un',0));
-template = mean(pulses,2);
-fprintf('done.\n');
-
-% handle diagnostic plots relevant to template building
-if options.diagnostics
-  t_template = (0:length(template)-1)'/data.header.sr;
-  t_pulses = repmat(t_template,1,length(pis)-2);
-  figure
-  plot(t_pulses,pulses,'--')
-  set(gca,'NextPlot','add')
-  plot(t_template,template,'k','lineWidth',3)
-  xlabel('time [s]')
-  ylabel('Amplitude')
-  title('Generated ppg template (bk) and pulses used (colored)')
+if ~isempty(options.missing)
+    [sts, missing] = pspm_get_timing('epochs', options.missing, 'seconds');
+    if sts < 1, return; end
+    index   = epochs2index(missing, numel(ppg), sr);
+    if (sum(index) > 0)
+        ppg(index) = 0;
+    end
 end
 
-%% Cross correlate the signal with the template and find peaks
-%--------------------------------------------------------------------------
-fprintf('Applying template.');
-ppg_corr = xcorr(data.data,template)/sum(template);
-% Truncate ppg_xcorr and realigne it so the max correlation corresponds to
-% templates peak and not beginning of template.
-ppg_corr = ppg_corr(length(data.data)-floor(.3*min_pulse_period):end-floor(.3*min_pulse_period));
-if options.diagnostics
-  t_ppg = (0:length(data.data)-1)'/data.header.sr;
-  figure
-  if length(t_ppg) ~= length(ppg_corr)
-    length(t_ppg)
-  end
-  plot(t_ppg,ppg_corr,t_ppg,data.data)
-  xlabel('time [s]')
-  ylabel('Amplitude')
-  title('ppg cross-corelated with template and ppg')
-  legend('ppg (X) template','ppg')
+% -------------------------------------------------------------------------
+%% Heartpy
+if strcmpi(options.method, 'heartpy')
+    % initialise python
+    if isempty(options.python_path)
+        sts = pspm_check_python;
+    else
+        sts = pspm_check_python(options.python_path);
+    end
+    if sts < 1, return; end
+    pspm_check_python_modules('heartpy');
+    
+    % prepare detection
+    ParametersSTFT = struct('TWindow', 16, ...
+        'TStep', 1, ...
+        'Tolerance', 0.5, ...
+        'HRCInit', 60);
+    TWindow = 16; % sliding window: 16s
+    TStep = 1; % moving step: 1s
+    Tolerance = 0.5;
+    HRCInit = 60; % init heart rate: 60bpm
+
+    filtered_ppg = py.heartpy.filter_signal(ppg, ...
+        pyargs('cutoff', [1,20], ...
+        'filtertype',  'bandpass', ...
+        'sample_rate', sr, ...
+        'order', 3));
+    filtered_ppg = double(py.array.array('d',(filtered_ppg)));
```

Comment:
I had some PPG data which is saved on my UCL one drive folder, but I have no access to it atm, apologies